File: app/src/main/java/androidx/room/RoomTrackingLiveData.java
/*
 * Copyright 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.room;


import android.annotation.SuppressLint;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;
import androidx.arch.core.executor.ArchTaskExecutor;
import androidx.lifecycle.LiveData;

import java.util.Set;
import java.util.concurrent.Callable;
import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

import eu.faircode.email.EntityLog;

/**
 * A LiveData implementation that closely works with {@link InvalidationTracker} to implement
 * database drive {@link androidx.lifecycle.LiveData} queries that are strongly hold as long
 * as they are active.
 * <p>
 * We need this extra handling for {@link androidx.lifecycle.LiveData} because when they are
 * observed forever, there is no {@link androidx.lifecycle.Lifecycle} that will keep them in
 * memory but they should stay. We cannot add-remove observer in {@link LiveData#onActive()},
 * {@link LiveData#onInactive()} because that would mean missing changes in between or doing an
 * extra query on every UI rotation.
 * <p>
 * This {@link LiveData} keeps a weak observer to the {@link InvalidationTracker} but it is hold
 * strongly by the {@link InvalidationTracker} as long as it is active.
 */
class RoomTrackingLiveData<T> extends LiveData<T> {
    @SuppressWarnings("WeakerAccess")
    final RoomDatabase mDatabase;

    @SuppressWarnings("WeakerAccess")
    final boolean mInTransaction;

    @SuppressWarnings("WeakerAccess")
    final Callable<T> mComputeFunction;

    private final InvalidationLiveDataContainer mContainer;

    @SuppressWarnings("WeakerAccess")
    final InvalidationTracker.Observer mObserver;

    final AtomicInteger queued = new AtomicInteger(0);
    final AtomicInteger running = new AtomicInteger(0);

    @SuppressWarnings("WeakerAccess")
    final AtomicBoolean mRegisteredObserver = new AtomicBoolean(false);

    @SuppressWarnings("WeakerAccess")
    final Runnable mRefreshRunnable = new Runnable() {
        @WorkerThread
        @Override
        public void run() {
            if (mRegisteredObserver.compareAndSet(false, true)) {
                mDatabase.getInvalidationTracker().addWeakObserver(mObserver);
            }
            try {
                running.incrementAndGet();

                T value = null;
                boolean computed = false;
                synchronized (mComputeFunction) {
                    int retry = 0;
                    while (!computed) {
                        try {
                            value = mComputeFunction.call();
                            computed = true;
                        } catch (Throwable e) {
                            if (++retry > 5) {
                                eu.faircode.email.Log.e(e);
                                break;
                            }
                            eu.faircode.email.Log.w(e);
                            try {
                                Thread.sleep(2000L);
                            } catch (InterruptedException ignored) {
                            }
                        }
                    }
                }
                if (computed) {
                    postValue(value);
                }
            } finally {
                queued.decrementAndGet();
                running.decrementAndGet();
            }
        }
    };

    @SuppressWarnings("WeakerAccess")
    final Runnable mInvalidationRunnable = new Runnable() {
        @MainThread
        @Override
        public void run() {
            if (running.get() == 0 && queued.get() > 0) {
                eu.faircode.email.Log.persist(EntityLog.Type.Debug,
                        mComputeFunction + " running=" + running + " queued=" + queued);
                return;
            }
            boolean isActive = hasActiveObservers();
            if (isActive) {
                queued.incrementAndGet();
                getQueryExecutor().execute(mRefreshRunnable);
            }
        }
    };

    @SuppressLint("RestrictedApi")
    RoomTrackingLiveData(
            RoomDatabase database,
            InvalidationLiveDataContainer container,
            boolean inTransaction,
            Callable<T> computeFunction,
            String[] tableNames) {
        mDatabase = database;
        mInTransaction = inTransaction;
        mComputeFunction = computeFunction;
        mContainer = container;
        mObserver = new InvalidationTracker.Observer(tableNames) {
            @Override
            public void onInvalidated(@NonNull Set<String> tables) {
                ArchTaskExecutor.getInstance().executeOnMainThread(mInvalidationRunnable);
            }
        };
    }

    @Override
    protected void onActive() {
        super.onActive();
        mContainer.onActive(this);
        getQueryExecutor().execute(mRefreshRunnable);
    }

    @Override
    protected void onInactive() {
        super.onInactive();
        mContainer.onInactive(this);
    }

    Executor getQueryExecutor() {
        if (mInTransaction) {
            return mDatabase.getTransactionExecutor();
        } else {
            return mDatabase.getQueryExecutor();
        }
    }
}

File: app/src/main/java/androidx/room/RoomTrackingLiveData.kt
/*
 * Copyright 2018 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package androidx.room

import android.annotation.SuppressLint
import androidx.arch.core.executor.ArchTaskExecutor
import androidx.lifecycle.LiveData
import java.lang.Exception
import java.lang.RuntimeException
import java.util.concurrent.Callable
import java.util.concurrent.Executor
import java.util.concurrent.atomic.AtomicBoolean

/**
 * A LiveData implementation that closely works with [InvalidationTracker] to implement
 * database drive [androidx.lifecycle.LiveData] queries that are strongly hold as long
 * as they are active.
 *
 * We need this extra handling for [androidx.lifecycle.LiveData] because when they are
 * observed forever, there is no [androidx.lifecycle.Lifecycle] that will keep them in
 * memory but they should stay. We cannot add-remove observer in [LiveData.onActive],
 * [LiveData.onInactive] because that would mean missing changes in between or doing an
 * extra query on every UI rotation.
 *
 * This [LiveData] keeps a weak observer to the [InvalidationTracker] but it is hold
 * strongly by the [InvalidationTracker] as long as it is active.
 */
@SuppressLint("RestrictedApi")
internal class RoomTrackingLiveData<T> (
    val database: RoomDatabase,
    private val container: InvalidationLiveDataContainer,
    val inTransaction: Boolean,
    val computeFunction: Callable<T?>,
    tableNames: Array<out String>
) : LiveData<T>() {
    val observer: InvalidationTracker.Observer = object : InvalidationTracker.Observer(tableNames) {
        override fun onInvalidated(tables: Set<String>) {
            ArchTaskExecutor.getInstance().executeOnMainThread(invalidationRunnable)
        }
    }
    val invalid = AtomicBoolean(true)
    val computing = AtomicBoolean(false)
    val registeredObserver = AtomicBoolean(false)
    val queued = eu.faircode.email.ObjectHolder<Int>(0)
    val lock = Object()
    val refreshRunnable = Runnable {
		eu.faircode.email.Log.i("MMM Refresh")
		synchronized(lock) {
			queued.value--
			if (queued.value < 0) {
				eu.faircode.email.Log.e("$computeFunction queued=$queued.value")
				queued.value = 0
			}
		}
		eu.faircode.email.Log.i("MMM Refreshing")

		if (registeredObserver.compareAndSet(false, true)) {
			database.invalidationTracker.addWeakObserver(observer)
		}

		var value: T? = null
		var computed = false
		synchronized(computeFunction) {
			var retry = 0
			while (!computed) {
				try {
					value = computeFunction.call()
					computed = true
				} catch (e: Throwable) {
					if (++retry > 5) {
						eu.faircode.email.Log.e(e)
						break
					}
					eu.faircode.email.Log.w(e)
					try {
						Thread.sleep(2000L)
					} catch (ignored: InterruptedException) {
					}
				}
			}
		}
		if (computed) {
			postValue(value)
		}
		eu.faircode.email.Log.i("MMM Refreshed")
	}

    val invalidationRunnable = Runnable {
		val isActive = hasActiveObservers()
		if (invalid.compareAndSet(false, true)) {
			if (isActive) {
			eu.faircode.email.Log.i("MMM Invalidate")
				synchronized(lock) {
					if (queued.value > 0) {
						eu.faircode.email.Log.persist(eu.faircode.email.EntityLog.Type.Debug, "$computeFunction queued=$queued.value")
					} else {
						queued.value++
						queryExecutor.execute(refreshRunnable)
					}
					eu.faircode.email.Log.i("MMM Invalidated")
				}
			}
		}
	}

    @Suppress("UNCHECKED_CAST")
    override fun onActive() {
		super.onActive()
		container.onActive(this as LiveData<Any>)
		eu.faircode.email.Log.i("MMM Active")
		synchronized(lock) {
			queued.value++
			queryExecutor.execute(refreshRunnable)
		}
		eu.faircode.email.Log.i("MMM Actived")
	}

    @Suppress("UNCHECKED_CAST")
    override fun onInactive() {
        super.onInactive()
        container.onInactive(this as LiveData<Any>)
    }

    val queryExecutor: Executor
        get() = if (inTransaction) {
            database.transactionExecutor
        } else {
            database.queryExecutor
        }
}
